### PR TITLE
Iterator specialization for EscapeDefaults

### PR DESCRIPTION
**Pull request by ticki**
Part of #30520. Completes #24214

### CODE REVIEW

File: src/libcore/char.rs [hunk 1]
```diff
@@ -537,4 +537,94 @@ impl Iterator for EscapeDefault {
...
+
+    fn nth(&mut self, n: usize) -> Option<char> {
+        let ch = match self.state {
+            EscapeDefaultState::Backslash(c)       => c,
+            EscapeDefaultState::Char(c)            => c,
+            EscapeDefaultState::Done               => return None,
+            EscapeDefaultState::Unicode(ref mut i) => return i.nth(n),
+        };
+
+        let start = if let Some(x) = self.get_offset() {
```
Comment (ranma42) [inline src/libcore/char.rs]: Wouldn't it be convenient to fold this into the previous match?
I did something similar here: https://github.com/ranma42/rust/commit/0cdfa6de5cf02cf9e8975deec58d72afe0221593

**Comment by ranma42** [comment]
Does this commit complete the tasks listed in #24214 ? AFAICT `EscapeUnicode` is still missing. I mentioned on the issue that have the branch https://github.com/ranma42/rust/tree/escape-iters for both `EscapeDefault` and `EscapeUnicode`, but there seemed to be no interest in merging it.

**Comment by Ticki** [comment]
@ranma42 Interesting. It's true that it doesn't fix the EscapeUnicode. Have you tried to send a pull request?

**Comment by ranma42** [comment]
I will send one soon. I had some unresolved questions, though, about how to handle code duplication and whether the methods should be `#[inline]`. I guess I will choose and fix as per suggestions on the PR.

**Comment by Ticki** [comment]
@ranma42 I don't think they should be so. They're not simple enough for inlining to be worth it.

**Comment by ranma42** [comment]
I believe that the test should be moved to `src/libcoretest/char.rs` (I tested my implementations using your test, too, but I was unable to get its results as part of "make check" until I did that ;) ).

File: src/libcore/char.rs [hunk 2]
```diff
@@ -537,4 +537,94 @@ impl Iterator for EscapeDefault {
...
+            x
+        } else {
+            return None;
+        };
+        let idx = start + n;
```
Comment (ranma42) [inline src/libcore/char.rs]: This sum can overflow if when `n = usize::MAX` and `start = 1` (i.e. `self.state` is `Char(_)`).

File: src/libcore/char.rs [hunk 3]
```diff
@@ -537,4 +537,94 @@ impl Iterator for EscapeDefault {
...
+        match self.state {
+            EscapeDefaultState::Unicode(iter)                              => iter.last(),
+            EscapeDefaultState::Done                                       => None,
+            EscapeDefaultState::Backslash(c) | EscapeDefaultState::Char(c) => Some(c),
+        }
+    }
+}
+
+#[test]
+fn ed_iterator_specializations() {
```
Comment (alexcrichton) [inline src/libcore/char.rs]: These tests will need to be moved to `src/libcoretests` as the unit tests of this crate aren't run.

File: src/libcore/char.rs [hunk 4]
```diff
@@ -537,4 +537,94 @@ impl Iterator for EscapeDefault {
...
+            0 => Some('\\'),
+            1 => Some(ch),
+            _ => None,
+        }
+    }
+
+    fn last(self) -> Option<char> {
+        match self.state {
+            EscapeDefaultState::Unicode(iter)                              => iter.last(),
+            EscapeDefaultState::Done                                       => None,
```
Comment (alexcrichton) [inline src/libcore/char.rs]: Stylistically we tend to not do this sort of alignment, so could you collapse the whitespace here?

File: src/libcore/char.rs (hunk 3 above)
Comment (alexcrichton) [inline src/libcore/char.rs]: Can you also be sure there are tests for `last`?

**Comment by Ticki** [comment]
@alexcrichton I have now fixed the comments you gave.

File: src/libcore/char.rs [hunk 5]
```diff
@@ -537,4 +537,61 @@ impl Iterator for EscapeDefault {
...
+
+    fn nth(&mut self, n: usize) -> Option<char> {
+        let ch = match self.state {
+            EscapeDefaultState::Backslash(c) => c,
+            EscapeDefaultState::Char(c) => c,
+            EscapeDefaultState::Done => return None,
+            EscapeDefaultState::Unicode(ref mut i) => return i.nth(n),
+        };
+
+        let start = if let Some(x) = self.get_offset() {
```
Comment (alexcrichton) [inline src/libcore/char.rs]: This method may not be necessary as it can just be folded into the `match` above

File: src/libcore/char.rs [hunk 6]
```diff
@@ -537,4 +537,61 @@ impl Iterator for EscapeDefault {
...
+        let idx = start + n;
+
+        // Update state
+        self.state = match idx {
+            0 => EscapeDefaultState::Char(ch),
+            _ => EscapeDefaultState::Done,
+        };
+
+        match idx {
+            0 => Some('\\'),
```
Comment (alexcrichton) [inline src/libcore/char.rs]: This logic may be simpler to just handle in the initial `match`, it's somewhat hard to follow what "offset" and "idx" are here

**Comment by Ticki** [comment]
@alexcrichton I updated it to reflect your points.

**Comment by ranma42** [comment]
@alexcrichton I have a branch which specialises `EscapeDefaults` and `EscapeUnicode` and uses the offset/index trick in both; the symmetry might make it more clear why it is convenient. I plan to make the PR after this is landed (to reuse some parts of this, including the new test).

**Comment by alexcrichton** [comment]
Ok, we can perhaps try to unify once they both exist, but at least for now this seems more understandable to me at least. 

@Ticki, can you also squash the commits together? Other than that looks good to me, thanks!

**Comment by Ticki** [comment]
@alexcrichton Yep. Will do.

**Comment by Ticki** [comment]
Squash done!

**Comment by alexcrichton** [comment]
Looks like some other snuck in by accident?

**Comment by Ticki** [comment]
Damn. Will fix.

**Comment by Ticki** [comment]
@alexcrichton Fixed.
